docs/design/kubeadm-api.md

```diff
@@ -0,0 +1,275 @@
+We want to identify the phases every single Kubernetes deployment goes through in order to get the cluster up and running.
+
+In this process, I've been thinking a lot about security and tried to put that first. Also, I've read the 
+kops, kargo and kube-up code in detail to see how they are doing these things. We don't want to make an API that's incompatible with those tools.
+
+So far (for kubeadm) I've identified these:
+
+#### Certificates
+
+First, certificates should be created in a directory.
+There should be:
+ - a CA certificate (`ca.pem`) with its private key (`ca-key.pem`)
+   - Is the public key necessary to have here (`ca-pub.pem`)?
+ - an API Server certificate (`apiserver.pem`) signed with `ca.pem` as the CA with its private key (`apiserver-key.pem`)
+ - a Service Account private key that's used for signing Service Accounts' tokens.
+   - TODO: Is the certificate for the serviceaccount private key necessary?
+
+Also, we should make sure that both RSA and ECDSA private keys work with kubeadm and have tests for both.
+We probably want to expose the private key bitsize also.
+
+One more option that could be useful to many would be to let the user specify their own ca.pem and ca-key.pem, 
+and kubeadm would just copy them over to the destination directory and proceed as usual. This might be beneficial to orgs
+that require a specific CA but lets the apiserver and serviceaccount keys/certs be generated by kubeadm.
+
+**Inputs**
+ - `AdvertiseAddresses` is needed for knowing which IPs the certs should be signed for
+ - `ExternalDNSNames` is needed for knowing which DNS names the certs should be signed for
+ - `DNSDomain` is needed for knowing which DNS name the internal kubernetes service has
+ - `ServiceSubnet` is needed for knowing which IP the internal kubernetes service is going to point to
+ - `CertificatesDir` is required for knowing where all certificates should be stored
+
+**Outputs**
+ - Files in directory `CertificatesDir`:
+   - `ca.pem`
+   - `ca-key.pem`
+   - `apiserver.pem`
+   - `apiserver-key.pem`
+   - ?`sa.pem`?
+   - `sa-key.pem`
+
+
+#### Generate KubeConfig files for the admin and the kubelet on the master
+
+The second phase takes the certificates generated in phase 1 and the public master endpoint as input and produces two kubeconfig files,
+`admin.conf` and `kubelet.conf` in `/etc/kubernetes/` by default.
+
+A general question here, do we want the KubernetesDir to be configurable at all or does it only add complexity/confusion if a user ever changed that path?
+I lean towards having that as a constant, since that seems to be the case with a lot of other deployments and the gain is very small.
+
+There also is the question whether this phase should generate both the kubelet and the admin config. Or should those be two different phases/steps?
+How can we make the kubelet have just a little permissions but still use it for bootstrapping the api server? Can start the master ("bootstrap") kubelet with full access, 
```

Comment:
Does bootkube do anything here? I agree that we want the kubelet to end up with "dropped permissions" even if it has to start out with root permissions to get things up and running.